### PR TITLE
coll: collective selection infrastructure (part 2)

### PR DESCRIPTION
**Pull request by AlexeyMalkhanov**
Introduced:
1) Backend decision tree implementation
2) Generic selection logic
3) New infrastructure for Bcast

https://anl.app.box.com/s/7ooyg3wbs5zdttmbtootws4xgjcy65p1

### CODE REVIEW

**Comment by AlexeyMalkhanov** [comment]
@raffenet Could you please take a look at the PR? It comprises collective selection backend based on a decision tree. As the first step, the full functionality has been implemented for Bcast.

File: src/util/coll_selection/coll_selection_tree_from_cvars.h [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+                                           (void *) &MPIDI_Bcast_inter_composition_alpha_cnt);
+
+    return bcast;
+}
+
+static MPIU_COLL_SELECTION_storage_handler
+MPIU_COLL_SELECTION_create_bcast_tree_topo_aware_compositions(MPIU_COLL_SELECTION_storage_handler
+                                                              parent, int coll_id)
+{
+/*  The tree which creates below stands for the following logic:
```
Comment (raffenet) [inline src/util/coll_selection/coll_selection_tree_from_cvars.h]: This where this PR loses me. I can't see a reason to include this in open-source MPICH unless there is greater interest from downstream. This relatively readable conditional is replaced by an abstract tree structure that will require a lot of effort to understand, validate, and/or modify.

Comment (raffenet) [reply src/util/coll_selection/coll_selection_tree_from_cvars.h]: At minimum, I'd like to hear from someone at Cray if they are interested in this. I'm not sure if our regular ch4 collaborators are on github, so I'll reach out by email.

**Comment by wesbland** [comment]
This effort is ongoing, but will be submitted as a new PR by @rkalidas

**Pull request closed without merging**